modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql-extensions/src/main/java/org/nuxeo/ecm/core/storage/sql/db/EmbeddedFunctions.java

```diff
@@ -132,82 +143,93 @@ public static boolean isAccessAllowed(Serializable id, Set<String> principals, S
                 }
-                if (first && newId == null) {
-                    // there is no parent for the first level
-                    // we may have a version on our hands, find the live doc
-                    try (PreparedStatement ps3 = conn.prepareStatement(
-                            "SELECT VERSIONABLEID FROM VERSIONS WHERE ID = ?")) {
-                        ps3.setObject(1, id);
-                        try (ResultSet rs = ps3.executeQuery()) {
-                            if (rs.next()) {
-                                newId = (Serializable) rs.getObject(1);
-                                if (rs.wasNull()) {
-                                    newId = null;
-                                }
-                            } else {
-                                // no such id
-                                newId = null;
-                            }
+                rowInfo = null;
+            } while (id != null);
+            // we've reached the root or a placeless document, deny access
+            return false;
+        }
+    }
+
+    protected static class RowInfo {
+
+        public Serializable parentId;
+
+        public boolean isVersion;
+
+        public Serializable versionableId;
+    }
+
+    protected static RowInfo getRowInfo(PreparedStatement psHier, PreparedStatement psVer, Serializable id)
+            throws SQLException {
+        psHier.setObject(1, id);
+        try (ResultSet rs = psHier.executeQuery()) {
+            RowInfo rowInfo = new RowInfo();
+            if (rs.next()) {
+                rowInfo.parentId = (Serializable) rs.getObject(1);
+                rowInfo.isVersion = rs.getBoolean(2);
+                if (rowInfo.isVersion) {
+                    psVer.setObject(1, id);
+                    try (ResultSet rs2 = psVer.executeQuery()) {
+                        if (rs2.next()) {
+                            rowInfo.versionableId = (Serializable) rs2.getObject(1);
                         }
                     }
                 }
-                first = false;
-                id = newId;
-            } while (id != null);
-            /*
-             * We reached the root, deny access.
-             */
-            if (isLogEnabled()) {
-                logDebug(" => false (root)");
             }
-            return false;
+            return rowInfo;
+        }
+    }
+
+    protected static Boolean getAccess(PreparedStatement psAcl, Serializable id, Set<String> principals,
+            Set<String> permissions) throws SQLException {
+        psAcl.setObject(1, id);
+        try (ResultSet rs = psAcl.executeQuery()) {
+            while (rs.next()) {
+                boolean grant = rs.getShort(1) != 0;
+                String permission = rs.getString(2);
+                String user = rs.getString(3);
+                if (principals.contains(user) && permissions.contains(permission)) {
+                    return grant;
+                }
```

Comment:
Less clarity but maybe this would avoid getting a boolean before knowing if it's aimed at the user regarding the permission ?
I tried adding a comment not to lose too much readability:
```suggestion
                String permission = rs.getString(2);
                String user = rs.getString(3);
                if (principals.contains(user) && permissions.contains(permission)) {
                    return rs.getShort(1) != 0; // the access grant boolean value
                }
```

Reply:
I prefer the clarity of getting all the values first, then acting on them. Also, this is just code that was moved.